File: app/src/main/java/com/rtyhdfgh/nrtdfgh/fragment/HomePageFragment.java
package com.rtyhdfgh.nrtdfgh.fragment;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.rtyhdfgh.nrtdfgh.R;
import com.rtyhdfgh.nrtdfgh.activity.ShowGoogsInfoActivity;
import com.rtyhdfgh.nrtdfgh.adapter.GoodsInfoItemAdapter;
import com.rtyhdfgh.nrtdfgh.entity.BaseEntity;
import com.rtyhdfgh.nrtdfgh.entity.GoodsEntity;
import com.rtyhdfgh.nrtdfgh.http.MainApi;
import com.rtyhdfgh.nrtdfgh.util.CommonUtil;
import com.rtyhdfgh.nrtdfgh.util.MyPreferences;
import com.trello.rxlifecycle2.components.support.RxFragment;

import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import retrofit2.Call;
import retrofit2.Callback;

public class HomePageFragment extends RxFragment {

    @BindView(R.id.goods_list)
    RecyclerView goodsList;
    @BindView(R.id.refresh_layout)
    SwipeRefreshLayout refreshLayout;
    @BindView(R.id.no_data_ll)
    View noDataLl;
    @BindView(R.id.parent_cl)
    View parentCl;
    @BindView(R.id.tishi_fl)
    View tishiFl;

    public View rootView;
    protected LayoutInflater layoutInflater;
    private Bundle bundle;
    private GoodsEntity goodsEntity;
    private int mobileType, type;
    private String phone;
    private GoodsInfoItemAdapter goodsInfoItemAdapter;

    public static HomePageFragment newInstance(int type) {
        Bundle args = new Bundle();
        args.putInt("type", type);
        HomePageFragment fragment = new HomePageFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        layoutInflater = inflater;
        if (rootView == null) {
            rootView = inflater.inflate(R.layout.fragment_home_page, null);
        } else {
            ViewGroup viewGroup = (ViewGroup) rootView.getParent();
            if (viewGroup != null) {
                viewGroup.removeView(rootView);
            }
        }
        ButterKnife.bind(this, rootView);
        return rootView;
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        Bundle arguments = getArguments();
        if (arguments != null) {
            type = arguments.getInt("type");
        }
        if (type == 2){
            tishiFl.setVisibility(View.GONE);
        } else {
            tishiFl.setVisibility(View.VISIBLE);
        }
        refreshLayout.setOnRefreshListener(() -> getGoodsList());
        parentCl.setOnClickListener(v -> productClick(goodsEntity));
        noDataLl.setOnClickListener(v -> getGoodsList());
    }

    @Override
    public void onResume() {
        super.onResume();
        getGoodsList();
    }

    private void getGoodsList() {
        mobileType = MyPreferences.getInt("mobileType");
        phone = MyPreferences.getString("phone");
        goodsEntity = null;
        MainApi.getRetrofitManager().getApiService().getGoodsList(mobileType, phone).enqueue(new Callback<BaseEntity<List<GoodsEntity>>>() {
            @Override
            public void onResponse(Call<BaseEntity<List<GoodsEntity>>> call, retrofit2.Response<BaseEntity<List<GoodsEntity>>> response) {
                if (response.body() == null){
                    noDataLl.setVisibility(View.VISIBLE);
                    goodsList.setVisibility(View.GONE);
                    return;
                }
                List<GoodsEntity> entity = response.body().getData();
                if (entity != null && entity.size() > 0){
                    noDataLl.setVisibility(View.GONE);
                    goodsList.setVisibility(View.VISIBLE);
                    goodsEntity = entity.get(0);
                    goodsInfoItemAdapter = new GoodsInfoItemAdapter(getActivity(), entity);
                    goodsInfoItemAdapter.setHasStableIds(true);
                    goodsInfoItemAdapter.setOnItemClickListener(new GoodsInfoItemAdapter.OnItemClickListener() {
                        @Override
                        public void itemClicked(GoodsEntity goodsEntity) {
                            productClick(goodsEntity);
                        }
                    });
                    goodsList.setHasFixedSize(true);
                    goodsList.setLayoutManager(new LinearLayoutManager(getActivity()));
                    goodsList.setAdapter(goodsInfoItemAdapter);
                } else {
                    noDataLl.setVisibility(View.VISIBLE);
                    goodsList.setVisibility(View.GONE);
                }
            }

            @Override
            public void onFailure(Call<BaseEntity<List<GoodsEntity>>> call, Throwable t) {

            }
        });
    }

    private void productClick(GoodsEntity goodsEntity){
        phone = MyPreferences.getString("phone");
        MainApi.getRetrofitManager().getApiService().productClick(goodsEntity.getId(), phone).enqueue(new Callback<BaseEntity>() {
            @Override
            public void onResponse(Call<BaseEntity> call, retrofit2.Response<BaseEntity> response) {
                bundle = new Bundle();
                bundle.putString("name", goodsEntity.getProductName());
                bundle.putString("url", goodsEntity.getUrl());
                CommonUtil.startActivity(getActivity(), ShowGoogsInfoActivity.class, bundle);
            }

            @Override
            public void onFailure(Call<BaseEntity> call, Throwable t) {
                bundle = new Bundle();
                bundle.putString("name", goodsEntity.getProductName());
                bundle.putString("url", goodsEntity.getUrl());
                CommonUtil.startActivity(getActivity(), ShowGoogsInfoActivity.class, bundle);
            }
        });
    }

}

File: app/src/main/java/com/asvsdfer/new_master_code_kotlin/ui/fragment/HomePageFragment.kt
package com.asvsdfer.new_master_code_kotlin.ui.fragment

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import com.asvsdfer.new_master_code_kotlin.R
import com.asvsdfer.new_master_code_kotlin.logic.model.BaseModel
import com.asvsdfer.new_master_code_kotlin.logic.model.EmptyModel
import com.asvsdfer.new_master_code_kotlin.logic.model.ProductModel
import com.asvsdfer.new_master_code_kotlin.logic.network.BaseNetWork
import com.asvsdfer.new_master_code_kotlin.ui.activity.GoodsDetailsActivity
import com.asvsdfer.new_master_code_kotlin.ui.adapter.GoodsListAdapter
import com.asvsdfer.new_master_code_kotlin.ui.livedata.MainViewModel
import com.asvsdfer.new_master_code_kotlin.util.PreferencesOpenUtil
import com.asvsdfer.new_master_code_kotlin.util.jumpActivity
import kotlinx.android.synthetic.main.fragment_home_page.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

class HomePageFragment : Fragment() {

    private val viewModel by lazy { ViewModelProvider(this).get(MainViewModel::class.java) }

    private var mobileType = 0
    private var type = 0
    private var phone: String? = null

    private var goodsListAdapter: GoodsListAdapter? = null
    private var productModel: ProductModel? = null

    companion object {
        fun newInstance(type: Int): HomePageFragment? {
            val args = Bundle()
            args.putInt("type", type)
            val fragment = HomePageFragment()
            fragment.arguments = args
            return fragment
        }
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_home_page, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        mobileType = PreferencesOpenUtil.getInt("mobileType")!!
        phone = PreferencesOpenUtil.getString("phone").toString()
        val arguments = arguments
        if (arguments != null) {
            type = arguments.getInt("type")
        }
        if (type == 1) {
            product_bg.visibility = View.GONE
            home_page_bg.visibility = View.VISIBLE
        } else {
            product_bg.visibility = View.VISIBLE
            home_page_bg.visibility = View.GONE
        }
        swipe_refresh_layout.setOnRefreshListener { getGoodsList() }
        root_ll.setOnClickListener {
            productModel?.let { productClick(it) }
        }
        no_data_ll.setOnClickListener { getGoodsList() }
        viewModel.getGoodsList.observe(viewLifecycleOwner) {
            swipe_refresh_layout.isRefreshing = false
            val model = it.getOrNull()
            if (model != null && model.isNotEmpty()) {
                no_data_ll.visibility = View.GONE
                goods_list.visibility = View.VISIBLE
                productModel = model[0]
                setListData(model)
            } else {
                no_data_ll.visibility = View.VISIBLE
                goods_list.visibility = View.GONE
            }
        }
    }

    override fun onResume() {
        super.onResume()
        getGoodsList()
    }

    private fun getGoodsList() {
        productModel = null
        phone?.let { viewModel.getGoodsList(mobileType, it) }
    }

    private fun setListData(mData: List<ProductModel>) {
        goodsListAdapter == null
        goodsListAdapter =
            activity?.let { GoodsListAdapter(it, mData) }
        goodsListAdapter?.setHasStableIds(true)
        goodsListAdapter?.setItemClickListener(object : GoodsListAdapter.ItemClickListener {
            override fun onItemClicked(item: ProductModel) {
                if (item != null) {
                    productClick(item)
                }
            }
        })
        goods_list.setHasFixedSize(true)
        goods_list.layoutManager = LinearLayoutManager(activity)
        goods_list.adapter = goodsListAdapter
    }

    private fun productClick(productModel: ProductModel) {
        BaseNetWork.placeService.productClick(productModel!!.id, phone!!)
            .enqueue(object : Callback<BaseModel<EmptyModel>> {
                override fun onFailure(call: Call<BaseModel<EmptyModel>>, t: Throwable) {
                    t.printStackTrace()
                }

                override fun onResponse(
                    call: Call<BaseModel<EmptyModel>>,
                    response: Response<BaseModel<EmptyModel>>
                ) {
                    activity?.let { it ->
                        jumpActivity<GoodsDetailsActivity>(it) {
                            putExtra("title", productModel.productName)
                            putExtra("url", productModel.url)
                        }
                    }
                }
            })
    }
}